ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -0,0 +1,225 @@
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.query.KafkaStreamsBuilder;
+import io.confluent.ksql.query.QueryError;
+import io.confluent.ksql.query.QueryErrorClassifier;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.rest.entity.StreamsTaskMetadata;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.LagInfo;
+import org.apache.kafka.streams.StreamsMetadata;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler;
+import org.apache.kafka.streams.processor.internals.namedtopology.KafkaStreamsNamedTopologyWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SharedKafkaStreamsRuntimeImpl implements SharedKafkaStreamsRuntime {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SharedKafkaStreamsRuntimeImpl.class);
+
+  private final KafkaStreamsBuilder kafkaStreamsBuilder;
+  private QueryErrorClassifier errorClassifier ;
+  private KafkaStreamsNamedTopologyWrapper kafkaStreams;
+  private ImmutableMap<String, Object> streamsProperties;
+  private final QueryMetadataImpl.TimeBoundedQueue queryErrors;
+  private final Map<String, PersistentQueriesInSharedRuntimesImpl> metadata;
+  public final Map<QueryId, Set<SourceName>> sources;
+
+  public SharedKafkaStreamsRuntimeImpl(final KafkaStreamsBuilder kafkaStreamsBuilder,
+                                       final int maxQueryErrorsQueueSize,
+                                       final Map<String, Object> streamsProperties) {
+    this.kafkaStreamsBuilder = kafkaStreamsBuilder;
+    kafkaStreams = kafkaStreamsBuilder.build(streamsProperties);
+    queryErrors
+        = new QueryMetadataImpl.TimeBoundedQueue(Duration.ofHours(1), maxQueryErrorsQueueSize);
+    this.streamsProperties = ImmutableMap.copyOf(streamsProperties);
+    metadata = new ConcurrentHashMap<>();
+    sources = new ConcurrentHashMap<>();
+    kafkaStreams.setUncaughtExceptionHandler(this::uncaughtHandler);
```

Comment:
Should we have a different exception handler which can tell which query inside the KS runtime was throwing the error, since I think it would be needed for the per-query healthcheck?

Reply:
I think it's still targeted for V1 to at least have the correct information about which query is affected. But not in this PR, we're _trying_ not to tackle every single thing at the same time 😜